src/GraphQL/Types/IAbstractGraphType.cs

```diff
@@ -0,0 +1,41 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace GraphQL.Types
+{
+    public interface IAbstractGraphType
+    {
+        Func<object, ObjectGraphType> ResolveType { get; set; }
+
+        List<ObjectGraphType> PossibleTypes { get; }        
```

Comment:
this is harder, The extension methods (below) needed a concrete implementation for adding types, I could just move that into each implementation directly though.